onnxruntime/core/providers/dml/DmlExecutionProvider/src/Operators/DmlDFT.h

```diff
@@ -0,0 +1,702 @@
+                index = Left + (index - 1) % range;
+            }
+            else if (index == End) {
+                index = Right + 1;
+            }
+            return index;
+        }
+    };
+    LoopRange loopRange_ = {};
+
+    struct DFTShaderConstants
+    {
+        uint32_t StartIndex;
+        uint32_t ElementCount;
+        uint32_t DFTIteration;
+        uint32_t IsInverse;
+        uint32_t InputSizes[4];
+        uint32_t InputStrides[4];
+        uint32_t OutputSizes[4];
+        uint32_t OutputStrides[4];
+        float Scale;
+    };
+
+public:
+    GpuDFTOperator(IMLOperatorKernelCreationContext* context)
+    {
+        ComPtr<IUnknown> executionObject;
+        context->GetExecutionInterface(executionObject.GetAddressOf());
+
+        ComPtr<ID3D12GraphicsCommandList> commandList;
+        executionObject.As(&commandList);
+
+        ORT_THROW_IF_FAILED(commandList->GetDevice(IID_ID3D12Device, &m_device));
+
+
+        ORT_THROW_IF_FAILED(context->GetAttribute("axis", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&axis_)));
+
+        int64_t inverse_i;
+        ORT_THROW_IF_FAILED(context->GetAttribute("inverse", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&inverse_i)));
+        inverse_ = static_cast<bool>(inverse_i);
+
+        int64_t onesided_i;
+        ORT_THROW_IF_FAILED(context->GetAttribute("onesided", MLOperatorAttributeType::Int, 1, sizeof(int64_t), reinterpret_cast<void*>(&onesided_i)));
+        is_onesided_ = static_cast<bool>(onesided_i);
+
+        ComPtr<IMLOperatorTensorShapeDescription> shapeDesc;
+        context->GetTensorShapeDescription(shapeDesc.GetAddressOf());
+
+        // Get the input and output shape sizes
+        uint32_t inputDimsSize;
+        ORT_THROW_IF_FAILED(shapeDesc->GetInputTensorDimensionCount(0, &inputDimsSize));
+        uint32_t outputDimsSize;
+        ORT_THROW_IF_FAILED(shapeDesc->GetOutputTensorDimensionCount(0, &outputDimsSize));
+        if (inputDimsSize != outputDimsSize)
+        {
+            ORT_THROW_IF_FAILED(E_FAIL);
+        }
+
+        // Get the input shape
+        inputDims_.resize(inputDimsSize);
+        ORT_THROW_IF_FAILED(shapeDesc->GetInputTensorShape(0, static_cast<uint32_t>(inputDims_.size()), inputDims_.data()));
+
+        // Get the output shape
+        outputDims_.resize(outputDimsSize);
+        ORT_THROW_IF_FAILED(shapeDesc->GetOutputTensorShape(0, static_cast<uint32_t>(outputDims_.size()), outputDims_.data()));
+
+        // For the number of total elements in the input and output shapes
+        outputDataSize_ = std::accumulate(outputDims_.begin(), outputDims_.end(), 1, std::multiplies<uint32_t>());
+        inputDataSize_ = std::accumulate(inputDims_.begin(), inputDims_.end(), 1, std::multiplies<uint32_t>());
+
+        // { before_dft_axis, axis, after_dft_axis, real_or_complex }
+        std::array<uint32_t, 4> reshapedInputSize = { 1, 1, 1, inputDims_.back() };
```

Comment:
>4 [](http://example.com/codeflow?start=29&length=1)

Are dimensions really limited to 4?

Per spec [https://github.com/onnx/onnx/blob/main/docs/Changelog.md#DFT-17](https://github.com/onnx/onnx/blob/main/docs/Changelog.md#DFT-17), it says it supports:

`[batch_idx][signal_dim1][signal_dim2]...[signal_dimN][1]`

Reply:
The shape can be reshaped to a 4 dimensional shape :)